src/parse.cpp

```diff
@@ -216,3 +216,127 @@ template <> double chemfiles::parse(string_view input) {
     // Return signed and scaled floating point result.
     return sign * (frac ? (value / scale) : (value * scale));
 }
+
+static const auto digits_upper = std::string("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ");
+static const auto digits_lower = std::string("0123456789abcdefghijklmnopqrstuvwxyz");
+
+static uint8_t digit_to_value(char c) {
+    if (c >= '0' && c <= '9') {
+        return c - '0';
+    }
+
+    if (c >= 'A' && c <= 'Z') {
+        return c - 'A' + 10;
+    }
+
+    // We know all the places this function is used, so this should be fine
+    return c - 'a' + 10;
+}
+
+static std::string encode_pure(const std::string& digits, uint64_t value) {
+    if (value == 0) {
+        return std::string(digits, 1);
+    }
+    
+    auto n = digits.length();
+    std::string result;
+    while (value != 0) {
+        auto rest = value / n;
+        result += digits[value - rest * n];
+        value = rest;
+    }
+
+    std::reverse(result.begin(), result.end());
+    return result;
+}
+
+static uint64_t decode_pure(string_view s) {
+    uint64_t result = 0;
+    auto n = digits_upper.length();
+    for (auto c : s) {
+        result *= n;
+        result += digit_to_value(c);
+    }
+    return result;
+}
+
+static int64_t pow_int(uint64_t base, uint64_t power) {
+    return static_cast<int64_t>(std::pow(base, power) + 0.5);
+}
+
+std::string chemfiles::encode_hydrid36(uint64_t width, int64_t value) {
+    
+    // the number is too negative to be encoded
+    if (value < (1 - pow_int(10, width - 1))) {
+        return std::string(width, '*');
+    }
+
+    // no need to encode
+    if (value < pow_int(10, width)) {
+        return std::to_string(value);
+    }
+    
+    // use upper case set
+    value -= pow_int(10, width);
+    if (value < 26 * pow_int(36, (width - 1))) {
+        value += 10 * pow_int(36, (width - 1));
+        return encode_pure(digits_upper, value);
+    }
+    
+    // use lower case set
+    value -= 26 * pow_int(36, width - 1);
+    if (value < 26 * pow_int(36, width - 1)) {
+        value += 10 * pow_int(36, width - 1);
+        return encode_pure(digits_lower, value);
+    }
+
+    // too large to be encoded
+    return std::string(width, '*');
+}
+
+int64_t chemfiles::decode_hybrid36(uint64_t width, string_view s) {
+
+    // This function is only called within chemfiles for fixed format files.
+    // Therefore, the width should also be the length of the string as this is
+    // known at compile time.
+    if (s.length() > width) {
+        throw error("the length of '{}' is greater than the width '{}', this is a bug in chemfiles", s, width);
+    }
+
+    auto f = s[0];
+    if (f == '-' || f == ' ' || std::isdigit(f)) {
```

Comment:
I try to not use `std::isdigit` or any functions from `<cctype>` or `<cclocale>` since they use a locale which might not correspond to the way we want to parse the values. 

There is a static `is_digit` function on the top of this file that you can use instead, and you can add `is_ascii_lower/is_ascii_upper` in the same way